File: Day1-9/7.py
import re
import itertools

abba_pat = re.compile(r'(.)((?!\1).)\2\1')


def get_aba(a):
    return [a[i:i+3] for i in range(len(a) - 2) if
            a[i] == a[i+2] and a[i] != a[i+1]]


def supports_tls(line):
    pat, found = re.compile(r'\[|\]'), False
    for idx, i in enumerate(pat.split(line)):
        res = abba_pat.search(i)
        if idx % 2 == 0 and res:
            found = True
        elif idx % 2 == 1 and res:
            return False
    return found


def supports_ssl(line):
    pat = re.compile(r'\[|\]')
    arr = pat.split(line)
    abas = list(itertools.chain.from_iterable(map(get_aba, arr[0::2])))
    babs = map(lambda i: '{0}{1}{0}'.format(i[1], i[0]), abas)
    return any(any(x in i for x in babs) for i in arr[1::2])


if __name__ == '__main__':
    assert supports_tls('abba[mnop]qrst')
    assert not supports_tls('abcd[bddb]xyyx')
    assert not supports_tls('aaaa[qwer]tyui')
    assert supports_tls('ioxxoj[asdfgh]zxcvbn')

    assert supports_ssl('aba[bab]xyz')
    assert not supports_ssl('xyx[xyx]xyx')
    assert supports_ssl('aaa[kek]eke')
    assert supports_ssl('zazbz[bzb]cdb')

    with open('7.txt', 'r') as f:
        lines = f.read().split()
        s1 = sum(1 for i in lines if supports_tls(i))
        s2 = sum(1 for i in lines if supports_ssl(i))
        print('Part 1: %s' % s1)
        print('Part 2: %s' % s2)

File: Day1-9/7.go
package main

import (
    "fmt"
    "strings"
    "io/ioutil"
)

func check(e error) {
    if e != nil {
        panic(e)
    }
}

func sslFields(r rune) bool {
    return r == '[' || r == ']'
}

func containsABBA(str string) bool {
    for i := 0; i < len(str) - 3; i++ {
        if(str[i] == str[i+3] && str[i+1] == str[i+2] && str[i] != str[i+1]) {
            return true
        }
    }
    return false
}

func supportsTLS(addr string) bool {
    chunks := strings.FieldsFunc(addr, sslFields)
    supportInside := false
    for i, chunk := range chunks {
        if i % 2 == 0 {
            // Not inside brackets
            if containsABBA(chunk) {
                supportInside = true
            }
        } else {
            // Inside brackets
            if containsABBA(chunk) {
                return false
            }
        }
    }
    return supportInside
}

func extractABAS(chunk string) []string {
    found := 0
    abas := make([]string, len(chunk))
    for i := 0; i < len(chunk) - 2; i++ {
        if chunk[i] == chunk[i+2] && chunk[i] != chunk[i+1] {
            abas[found] = chunk[i:i+3]
            found++
        }
    }
    return abas[:found]
}

func supportsSSL(addr string) bool {
    chunks := strings.FieldsFunc(addr, sslFields)
    abas := make([]string, 0)
    babs := make(map[string]bool)
    for i, chunk := range chunks {
        if i % 2 == 0 {
            // Not inside brackets
            abas = append(abas, extractABAS(chunk)...)
        } else {
            // Inside brackets
            for _, bab := range extractABAS(chunk) {
                babs[bab] = true
            }
        }
    }
    for _, aba := range abas {
        corresponding := fmt.Sprintf("%c%c%c", aba[1], aba[0], aba[1])
        if _, ok := babs[corresponding]; ok {
            return true
        }
    }
    return false
}

func part1(addresses []string) int {
    c := 0
    for _, str := range addresses {
        if(supportsTLS(str)) {
            c++
        }
    }
    return c
}

func part2(addresses []string) int {
    c := 0
    for _, str := range addresses {
        if(supportsSSL(str)) {
            c++
        }
    }
    return c
}

func main() {
    dat, err := ioutil.ReadFile("./7.txt")
    check(err)
    lines := strings.Split(string(dat), "\n")
    fmt.Printf("Part 1: %d\n", part1(lines))
    fmt.Printf("Part 2: %d\n", part2(lines))
}